pybind11 type_caster for at::Generator and custom RNG python test

pbelevich opened the pull request:
This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35167 Move normal() to DistributionTemplates
* **#34774 pybind11 type_caster for at::Generator and custom RNG python test**

Differential Revision: [D20549451](https://our.internmc.facebook.com/intern/diff/D20549451)

pbelevich committed:
pybind11 type_caster for at::Generator and custom RNG python test

[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] pybind11 type_caster for at::Generator and custom RNG python test"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] pybind11 type_caster for at::Generator and custom RNG python test"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] pybind11 type_caster for at::Generator and custom RNG python test"




[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`

* #34468 [WIP] Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>

[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`



[ghstack-poisoned]

ezyang commented on torch/csrc/Generator.cpp:
```diff
@@ -221,3 +221,38 @@ bool THPGenerator_init(PyObject *module)
+  return THPGenerator_NewWithVar((PyTypeObject *)THPGeneratorClass, std::move(gen));
+}
+
+PyObject* THPGenerator_NewWithVar(PyTypeObject* type, Generator gen)
```
The place you copy pasted this from has a comment talking about what invariant should be upheld before calling this function; that comment should be preserved here!

ezyang replied:
don't forget about this comment!

pbelevich replied:
I added a comment in the header, but I will add it here as well

ezyang commented:
General comment: I know there is existing Python binding code for Generator, and I don't see that code being adjusted in this PR. So I feel there is definitely something missing here.

ezyang commented:
Destructor for THPGenerator needs to clear the pyobj field (otherwise it's dangling). Needs more tests about preserving object identity.

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`



[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`



[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`



[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`



[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`



[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`



[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`



[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`



[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`



[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`



[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`



[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`


Differential Revision: [D20549451](https://our.internmc.facebook.com/intern/diff/D20549451)

[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`


Differential Revision: [D20549451](https://our.internmc.facebook.com/intern/diff/D20549451)

[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`


Differential Revision: [D20549451](https://our.internmc.facebook.com/intern/diff/D20549451)

[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`


Differential Revision: [D20549451](https://our.internmc.facebook.com/intern/diff/D20549451)

[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`


Differential Revision: [D20549451](https://our.internmc.facebook.com/intern/diff/D20549451)

[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`


Differential Revision: [D20549451](https://our.internmc.facebook.com/intern/diff/D20549451)

[ghstack-poisoned]

pbelevich committed:
Update on "pybind11 type_caster for at::Generator and custom RNG python test"

This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`


Differential Revision: [D20549451](https://our.internmc.facebook.com/intern/diff/D20549451)

[ghstack-poisoned]
